### PR TITLE
[WFCORE-433]: Using Git for persisting configuration files

### PR DESCRIPTION
**Pull request by ehsavoie**
Jira: https://issues.jboss.org/browse/WFCORE-433
JBEAP: https://issues.jboss.org/browse/EAP7-728
PR: https://github.com/wildfly/wildfly-core/pull/3352

### CODE REVIEW

File: controller/WFCORE-433_git-persistence.adoc [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+:email:             ehugonne@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org
+
+== Overview
+
+Replace the current persistence mechanism for configuration files and history by one using a real version control system.
```
Comment (dmlloyd) [inline controller/WFCORE-433_git-persistence.adoc]: Do you really mean "replace" or will the new system supplement the existing one?

Comment (ehsavoie) [reply controller/WFCORE-433_git-persistence.adoc]: Supplement :)

**Comment by kabir** [comment]
@ehsavoie Can you please add in the new Community Documentation section I just mentioned on wildfly-dev?

File: controller/WFCORE-433_git-persistence.adoc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+These items are not required but would be nice to have if possible.
+* Support for branch selection, but you can’t change the branch being used once the server has started-up.
+* Support for tag selection, but selecting a tag means that we are in a read-only mode.
+* Support for automatic conflict resolution with strategies.
+* Support for Git Large File Storage https://git-lfs.github.com/[https://git-lfs.github.com/]
+
+=== Non-Requirements
+
+These are items that are explicitly not required.
+* Managing conflicts thought the management API / HAL / CLI
```
Comment (jmesnil) [inline controller/WFCORE-433_git-persistence.adoc]: [minor] the list is not displayed correctly (needs a newline before

File: controller/WFCORE-433_git-persistence.adoc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+* Extension to more elements of the configuration (aka managed contents) being versioned. When a managed content is no longer relevant it should be removed from the repository. +
+This supposes that the directory structure of the content directory and the configuration directory are the same as in the default configuration. Changing the jboss.server.base.dir is not a problem but we can’t manage several different places on a filesystem through Git.
+* Integration with Elytron for authentication : as we might need to authenticate on the remote repository instead of passing the username/password in clear text we could be using Elytron and pass the authentication credentials through a wildfly-config.xml.
+* Conflicts are to be resolved automatically. If that fails then it is up to the user to clean the mess using git itself as the tool is not meant to provide a full git implementation in jboss-cli.
+
+=== Nice-to-Have Requirements
+
+These items are not required but would be nice to have if possible.
+* Support for branch selection, but you can’t change the branch being used once the server has started-up.
```
Comment (jmesnil) [inline controller/WFCORE-433_git-persistence.adoc]: [minor] the list is not displayed correctly (needs a newline before

File: controller/WFCORE-433_git-persistence.adoc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+This supposes that the directory structure of the content directory and the configuration directory are the same as in the default configuration. Changing the jboss.server.base.dir is not a problem but we can’t manage several different places on a filesystem through Git.
+* Integration with Elytron for authentication : as we might need to authenticate on the remote repository instead of passing the username/password in clear text we could be using Elytron and pass the authentication credentials through a wildfly-config.xml.
+* Conflicts are to be resolved automatically. If that fails then it is up to the user to clean the mess using git itself as the tool is not meant to provide a full git implementation in jboss-cli.
+
+=== Nice-to-Have Requirements
+These items are not required but would be nice to have if possible:
+
+* Support for branch selection, but you can’t change the branch being used once the server has started-up.
+* Support for tag selection, but selecting a tag means that we are in a read-only mode.
```
Comment (nziakova) [inline controller/WFCORE-433_git-persistence.adoc]: In current implementation new config changes can be committed (results in detached state). I think it's OK, just needs a note in the docs.

**Pull request merged**